enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,198 @@
+
+PROs:
+
+- advanced plans to provide a mutation webhook implementing a MutationTemplate (design doc [here|https://docs.google.com/document/d/1G7WgZKx1Y3VOTUjrqn7DjDaZgSKCIZowILm_I6psrw0/edit?pli=1#
+])
+- seem to be interested in adding a updates to existing resources in reaction to MutationTemplate changes (but no hard commitment on this)
+- a regularly release opensource project with  community backed by companies like google, ms, RH, ...
+
+CONs:
+- don't provide any functionality we need yet
+- reactive resource updates still uncertain
+- alpha version of the mutation webhook only available by the end of the year
+- an overkill for our immediate needs, provides a much broader functionality than we need
+
+### In-house solution
+
+If using gatekeeper is not feasible, a simple custom solution should be considered. Below is a proposal of such a project.
+
+PROs:
+
+- we own it, can add anything that is needed
+
+CONs:
+
+- still vaporware
+
+A mutating webhook (plus a  controller) will inspect created resources in specified namespaces for labeling scheme compliance. The webhook will check the namespace, type, ... of the created resource, match it to the configured labeling scheme, and add any missing labels/annotations. The webhook could also update incorrectly set labels/annotations if requested.
+
+The webhook will be configured by means of a LabelSchemeConfig CRD. The LabelSchemeConfig CRD will specify a target namespace, allowing to specify a label scheme for the specific namespace (an alternative solution of having the LabelSchemeConfig CRD namespaced would allow project admins to access the resource, which is not desirable).  
+ The controller will look for LabelSchemeConfig CRDs in it's own namespace.
+
+A sample LabelSchemeConfig resource is shown below:
+
+```yaml
+apiVersion: v1
+kind: LabelSchemeConfig
+metadata:
+ name: labeling-scheme-config-for-namespace1
+ namespace: controller-namespace
+spec:
+ labelSchemeNamespace: namespace1
+ labelScheme:
+   - apiVersion: networking.k8s.io/v1/
+     kind: NetworkPolicy
+     labelSelector: ""
+     labels:
+       label1-key: label1-value
+       label2-key: label2-value
+     annotations:
+       annotations1-key: annotations1-value
+   - apiVersion: v1
+     kind: Pod
+     labelSelector:
+       - app: custom-app
+     resourceName: resource-name
+     labels:
+       - label1-key: label1-value
+       - label2-key: label2-value
+     annotations:
+       annotations1-key: annotations1-value
+```
+
+The LabelSchemeConfig CRD will require a separate resource for each namespace, which although elegant migth be laborsome to maintain. An alternative solution could be to allow to specify a list of LabelSchemeConfig's in one resource:
+
+```yaml
+apiVersion: v1
+kind: LabelSchemeConfig
+metadata:
+ name: labeling-scheme-config-for-multiple-namespaces
+ namespace: controller-namespace
+spec:
```

Comment:
a spec as slice is uncommon

Reply:
Also note that you cannot define polymorphic types. Either spec is an object or it is a slice. It cannot be both.

Reply:
Added "schemes" slice to spec.
Another option would be to have multiple LabelSchemeConfig's with a single scheme inside it.